File: Libs/pollilib/javascript/polliLib/feeds.js
export const FeedsMixin = (Base) => class extends Base {
  async *image_feed_stream(options = {}) {
    const {
      referrer = null,
      token = null,
      timeoutMs,
      reconnect = false,
      retryDelayMs = 10_000,
      yieldRawEvents = false,
      includeBytes = false,
      includeDataUrl = false,
    } = options;
    const feedUrl = new URL('https://image.pollinations.ai/feed');
    if (referrer) feedUrl.searchParams.set('referrer', referrer);
    if (token) feedUrl.searchParams.set('token', token);
    const limit = this._resolveTimeout(timeoutMs, 300_000);

    const connect = async function* (self) {
      const controller = new AbortController();
      const t = setTimeout(() => controller.abort(), limit);
      try {
        const resp = await self.fetch(feedUrl, { method: 'GET', headers: { 'Accept': 'text/event-stream' }, signal: controller.signal });
        if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
        for await (const line of iterateSSELines(resp)) {
          if (!line.startsWith('data:')) continue;
          const data = line.slice(5).trim();
          if (data === '[DONE]') break;
          if (yieldRawEvents) { yield data; continue; }
          try {
            const ev = JSON.parse(data);
            if (includeBytes || includeDataUrl) {
              const imgUrl = ev.imageURL || ev.image_url;
              if (imgUrl) {
                const r = await self.fetch(imgUrl);
                if (r.ok) {
                  const buf = Buffer.from(await r.arrayBuffer());
                  if (includeDataUrl) {
                    const ctype = r.headers?.get ? (r.headers.get('Content-Type') || 'image/jpeg') : (r.headers?.['Content-Type'] || 'image/jpeg');
                    const b64 = buf.toString('base64');
                    ev.image_data_url = `data:${ctype};base64,${b64}`;
                  } else if (includeBytes) {
                    ev.image_bytes = buf;
                  }
                }
              }
            }
            yield ev;
          } catch { /* ignore malformed */ }
        }
      } finally { clearTimeout(t); }
    };

    if (!reconnect) {
      yield* (await connect(this));
      return;
    }
    for (;;) {
      try {
        yield* (await connect(this));
      } catch { /* ignore */ }
      await new Promise((r) => setTimeout(r, retryDelayMs));
    }
  }

  async *text_feed_stream(options = {}) {
    const {
      referrer = null,
      token = null,
      timeoutMs,
      reconnect = false,
      retryDelayMs = 10_000,
      yieldRawEvents = false,
    } = options;
    const feedUrl = new URL('https://text.pollinations.ai/feed');
    if (referrer) feedUrl.searchParams.set('referrer', referrer);
    if (token) feedUrl.searchParams.set('token', token);
    const limit = this._resolveTimeout(timeoutMs, 300_000);
    const connect = async function* (self) {
      const controller = new AbortController();
      const t = setTimeout(() => controller.abort(), limit);
      try {
        const resp = await self.fetch(feedUrl, { method: 'GET', headers: { 'Accept': 'text/event-stream' }, signal: controller.signal });
        if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
        for await (const line of iterateSSELines(resp)) {
          if (!line.startsWith('data:')) continue;
          const data = line.slice(5).trim();
          if (data === '[DONE]') break;
          if (yieldRawEvents) { yield data; continue; }
          try { yield JSON.parse(data); } catch { }
        }
      } finally { clearTimeout(t); }
    };
    if (!reconnect) { yield* (await connect(this)); return; }
    for (;;) {
      try { yield* (await connect(this)); } catch {}
      await new Promise((r) => setTimeout(r, retryDelayMs));
    }
  }
};

async function *iterateSSELines(resp) {
  if (resp.body && typeof resp.body.getReader === 'function') {
    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buf = '';
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      const parts = buf.split(/\r?\n/);
      buf = parts.pop() ?? '';
      for (const l of parts) yield l;
    }
    if (buf) yield buf;
    return;
  }
  const text = await resp.text();
  for (const line of String(text).split(/\r?\n/)) yield line;
}

File: Libs/pollilib/python/polliLib/feeds.py
from __future__ import annotations

from typing import Any, Dict, Iterator, Optional


class FeedsMixin:
    def image_feed_stream(
        self,
        *,
        referrer: Optional[str] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = None,
        reconnect: bool = False,
        retry_delay: float = 10.0,
        yield_raw_events: bool = False,
        include_bytes: bool = False,
        include_data_url: bool = False,
    ) -> Iterator[Any]:
        """
        Stream the public image feed via SSE.
        - Yields dicts or raw JSON strings when yield_raw_events=True.
        - include_bytes -> add 'image_bytes' to each dict
        - include_data_url -> add 'image_data_url' (base64) to each dict
        """
        feed_url = "https://image.pollinations.ai/feed"

        eff_timeout = self._resolve_timeout(timeout, 300.0)

        def _connect() -> Iterator[Any]:
            params: Dict[str, Any] = {}
            if referrer:
                params["referrer"] = referrer
            if token:
                params["token"] = token
            headers = {"Accept": "text/event-stream"}
            with self.session.get(feed_url, params=params, headers=headers, stream=True, timeout=eff_timeout) as resp:
                resp.raise_for_status()
                for raw in resp.iter_lines(decode_unicode=True):
                    if not raw:
                        continue
                    if isinstance(raw, bytes):
                        try:
                            raw = raw.decode("utf-8", errors="ignore")
                        except Exception:
                            continue
                    line = raw.strip()
                    if not line or line.startswith(":"):
                        continue
                    if not line.startswith("data:"):
                        continue
                    data = line[len("data:"):].strip()
                    if data == "[DONE]":
                        break
                    if yield_raw_events:
                        yield data
                        continue
                    try:
                        import json as _json, base64 as _b64
                        ev = _json.loads(data)
                        if include_data_url or include_bytes:
                            img_url = ev.get("imageURL") or ev.get("image_url")
                            if img_url:
                                r = self.session.get(img_url, timeout=eff_timeout)
                                r.raise_for_status()
                                content = r.content
                                if include_data_url:
                                    ctype = r.headers.get("Content-Type", "image/jpeg")
                                    b64 = _b64.b64encode(content).decode("utf-8")
                                    ev["image_data_url"] = f"data:{ctype};base64,{b64}"
                                elif include_bytes:
                                    ev["image_bytes"] = content
                        yield ev
                    except Exception:
                        continue

        if not reconnect:
            yield from _connect()
            return

        import time as _time
        while True:
            try:
                for item in _connect():
                    yield item
            except Exception:
                pass
            _time.sleep(retry_delay)

    def text_feed_stream(
        self,
        *,
        referrer: Optional[str] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = None,
        reconnect: bool = False,
        retry_delay: float = 10.0,
        yield_raw_events: bool = False,
    ) -> Iterator[Any]:
        feed_url = "https://text.pollinations.ai/feed"

        eff_timeout = self._resolve_timeout(timeout, 300.0)

        def _connect() -> Iterator[Any]:
            params: Dict[str, Any] = {}
            if referrer:
                params["referrer"] = referrer
            if token:
                params["token"] = token
            headers = {"Accept": "text/event-stream"}
            with self.session.get(feed_url, params=params, headers=headers, stream=True, timeout=eff_timeout) as resp:
                resp.raise_for_status()
                for raw in resp.iter_lines(decode_unicode=True):
                    if not raw:
                        continue
                    if isinstance(raw, bytes):
                        try:
                            raw = raw.decode("utf-8", errors="ignore")
                        except Exception:
                            continue
                    line = raw.strip()
                    if not line or line.startswith(":"):
                        continue
                    if not line.startswith("data:"):
                        continue
                    data = line[len("data:"):].strip()
                    if data == "[DONE]":
                        break
                    if yield_raw_events:
                        yield data
                        continue
                    try:
                        import json as _json
                        yield _json.loads(data)
                    except Exception:
                        continue

        if not reconnect:
            yield from _connect()
            return

        import time as _time
        while True:
            try:
                for item in _connect():
                    yield item
            except Exception:
                pass
            _time.sleep(retry_delay)

